lib/Common/DataStructures/WeakReferenceDictionary.h

```diff
@@ -44,4 +44,668 @@ namespace JsUtil
 
         PREVENT_COPY(WeakReferenceDictionary);
     };
+
+
+    template <class TKey>
+    class WeakRefRegionValueDictionaryEntry : public SimpleDictionaryKeyEntry<TKey>
+    {
+    public:
+        void Clear()
+        {
+            this->value = TKey();
+        }
+    };
+
+    template <
+        class TKey,
+        class TValue,
+        class SizePolicy = PowerOf2SizePolicy,
+        template <typename ValueOrKey> class Comparer = DefaultComparer,
+        typename Lock = NoResizeLock
+    >
+    class WeakReferenceRegionDictionary : protected Lock, public IWeakReferenceDictionary
```

Comment:
This class is largely a copy of `BaseDictionary` with a few changes: it only uses the Recycler rather than allowing arbitrary allocators, and instead of having a single list `entries` of key/value/next tuples, it has one list of `key/next` pairs, and a separate list of `values` which is allocated in weakly-referencing Recycler memory.